File: day22/Main.hs
import           Prelude                 hiding ( round )
import           Data.Set                       ( Set )
import qualified Data.Set                      as Set
import           Text.Parsec
import           Text.Parsec.String             ( Parser )

type Card = Int
type Deck = [Card]

data Winner = P1 | P2

parseCard :: Parser Card
parseCard = read <$> (many1 digit <* newline)

parsePlayer :: Parser Deck
parsePlayer = do
  string "Player "
  digit
  char ':'
  newline
  many1 parseCard

parse2Players :: Parser (Deck, Deck)
parse2Players = do
  one <- parsePlayer
  newline
  two <- parsePlayer
  return (one, two)

score :: Deck -> Int
score deck = sum $ zipWith (*) [1 ..] $ reverse deck

scoreWinner :: (Winner, Deck, Deck) -> Int
scoreWinner (P1, p1, _ ) = score p1
scoreWinner (P2, _ , p2) = score p2

draw :: Deck -> (Card, Deck)
draw []           = error "Cannot draw from an empty deck!"
draw (top : rest) = (top, rest)

collect :: (Card, Card) -> Deck -> Deck
collect (c1, c2) d = d ++ [c1, c2]

round :: (Deck, Deck) -> (Deck, Deck)
round (p1, p2) =
  let ((top1, rest1), (top2, rest2)) = (draw p1, draw p2)
      res | top1 > top2 = (collect (top1, top2) rest1, rest2)
          | top1 < top2 = (rest1, collect (top2, top1) rest2)
          | otherwise   = error "Cards are equal!"
  in  res

combat :: (Deck, Deck) -> (Winner, Deck, Deck)
combat (p1, []) = (P1, p1, [])
combat ([], p2) = (P2, [], p2)
combat decks    = combat $ round decks

roundR :: (Deck, Deck) -> (Deck, Deck)
roundR (p1, p2) =
  let ((top1, rest1), (top2, rest2)) = (draw p1, draw p2)
      win1                           = (collect (top1, top2) rest1, rest2)
      win2                           = (rest1, collect (top2, top1) rest2)
      res
        | top1 <= length rest1 && top2 <= length rest2
        = let rDeck1 = take top1 rest1
              rDeck2 = take top2 rest2
          in  case combatR (rDeck1, rDeck2) of
                (P1, _, _) -> win1
                (P2, _, _) -> win2
        | top1 > top2
        = win1
        | top1 < top2
        = win2
        | otherwise
        = error "Cards are equal!"
  in  res

combatR' :: Set (Deck, Deck) -> (Deck, Deck) -> (Winner, Deck, Deck)
combatR' _ (p1, []) = (P1, p1, [])
combatR' _ ([], p2) = (P2, [], p2)
combatR' seen decks@(d1, d2)
  | Set.member decks seen = (P1, d1, d2)
  | otherwise             = combatR' (Set.insert decks seen) (roundR decks)

combatR :: (Deck, Deck) -> (Winner, Deck, Deck)
combatR = combatR' Set.empty

main = do
  input <- readFile "input.txt"
  let Right (deck1, deck2) = parse parse2Players "" input

  putStrLn "Part 1:"
  print $ scoreWinner $ combat (deck1, deck2)

  putStrLn "Part 2:"
  print $ scoreWinner $ combatR (deck1, deck2)

File: day22/main.go
package main

import (
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
)

func max(a, b int) int {
	if a < b {
		return b
	}
	return a
}

func check(e error) {
	if e != nil {
		panic(e)
	}
}

func parseDeck(lines []string) deck {
	cards := []int{}
	for _, line := range lines {
		if line == "" {
			continue
		}
		c, err := strconv.Atoi(line)
		check(err)
		cards = append(cards, c)
	}
	return deck{cards}
}

func parse2players(input string) (deck, deck) {
	players := strings.Split(input, "\n\n")
	deck1, deck2 := strings.Split(players[0], "\n"), strings.Split(players[1], "\n")
	p1 := parseDeck(deck1[1:])
	p2 := parseDeck(deck2[1:])
	return p1, p2
}

type deck struct {
	cards []int
}

func (d *deck) draw() int {
	top := d.cards[0]
	d.cards = d.cards[1:]
	return top
}

func (d *deck) add(card int) *deck {
	d.cards = append(d.cards, card)
	return d
}

func (d *deck) score() int {
	l := len(d.cards)
	total := 0
	for i, mult := 0, l; i < l; i, mult = i+1, mult-1 {
		total += mult * d.cards[i]
	}
	return total
}

func (d *deck) size() int {
	return len(d.cards)
}

func (d *deck) empty() bool {
	return d.size() == 0
}

func (d *deck) pr() string {
	return fmt.Sprint(d.cards)
}

func (d *deck) take(n int) deck {
	cp := make([]int, n)
	copy(cp, d.cards[:n])
	return deck{cards: cp}
}

func score(winner int, deck1, deck2 deck) int {
	switch winner {
	case 1:
		return deck1.score()
	case 2:
		return deck2.score()
	}
	panic("Winner ID not valid!")
}

func combat(deck1, deck2 deck) ( /* winner */ int, deck, deck) {
	for !deck1.empty() && !deck2.empty() {
		top1, top2 := deck1.draw(), deck2.draw()
		if top1 > top2 {
			deck1.add(top1).add(top2)
		} else {
			deck2.add(top2).add(top1)
		}
	}
	if deck1.empty() {
		return 2, deck1, deck2
	}
	return 1, deck1, deck2
}

func hashDecks(deck1, deck2 deck) string {
	return deck1.pr() + "|" + deck2.pr()
}

func combatRecursive(deck1, deck2 deck) ( /* winner */ int, deck, deck) {
	seen := map[string]bool{}
	for !deck1.empty() && !deck2.empty() {
		roundHash := hashDecks(deck1, deck2)
		if seen[roundHash] {
			return 1, deck1, deck2
		}
		seen[roundHash] = true

		top1, top2 := deck1.draw(), deck2.draw()

		if deck1.size() >= top1 && deck2.size() >= top2 {
			winner, _, _ := combatRecursive(deck1.take(top1), deck2.take(top2))
			switch winner {
			case 1:
				deck1.add(top1).add(top2)
			case 2:
				deck2.add(top2).add(top1)
			}
		} else if top1 > top2 {
			deck1.add(top1).add(top2)
		} else {
			deck2.add(top2).add(top1)
		}
	}
	if deck1.empty() {
		return 2, deck1, deck2
	}
	return 1, deck1, deck2
}

func main() {
	fi, err := ioutil.ReadFile("day22/input.txt")
	check(err)
	ficontent := string(fi)

	{
		p1, p2 := parse2players(ficontent)

		fmt.Println("Part 1:")
		fmt.Println(score(combat(p1, p2)))
	}

	{
		p1, p2 := parse2players(ficontent)

		fmt.Println("Part 2:")
		fmt.Println(score(combatRecursive(p1, p2)))
	}
}
